backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/ModularContractProvider.java

```diff
@@ -219,4 +220,41 @@ private Optional<FlatrateTermId> extractFlatrateTermId(@NonNull final I_C_Invoic
 
 		return Optional.empty();
 	}
+
+	@NonNull
+	public Stream<FlatrateTermId> streamModularPurchaseContractsForShipmentLine(@NonNull final InOutLineId inOutLineId)
+	{
+		final I_M_InOutLine inOutLineRecord = inOutDAO.getLineByIdInTrx(inOutLineId);
+		final I_M_InOut inOutRecord = inOutDAO.getById(InOutId.ofRepoId(inOutLineRecord.getM_InOut_ID()));
+		if (!inOutRecord.isSOTrx() || inOutLineRecord.getMovementQty().signum() < 0)
+		{
+			return Stream.empty();
+		}
+
+		final I_C_Order order = orderBL.getById(OrderId.ofRepoId(inOutLineRecord.getC_Order_ID()));
+
+		final WarehouseId warehouseId = WarehouseId.ofRepoId(order.getM_Warehouse_ID()); // C_Order.M_Warehouse_ID is mandatory and warehouseBL.getBPartnerId demands NonNull
```

Comment:
why not using the warehouse of InOut?

Reply:
I need to check the old issue again for this. Related to using linked **Purchase Modular Contract** on shipment here I think